### ISSUE TITLE
How to create & manage individual db for each user?

### ISSUE THREAD

**Issue by rkanani108** [litedb-org/LiteDB#1849]
Hi @mbdavid ,

I am planning to use Litedb in our product. In that, we want to create a database for each individual user and store user-specific information to the respected database.

I just want to confirm below things.

1.  In our use-case Litedb is feasible or not.
2. Is there any limit to store an amount of data for each individual db?
3. Let me provide if any sample code is available to fit with our use-case.

Thanks'
Rakesh Kanani

**Comment by AlBannaTechno** [comment]
@rkanani108 Hi, I know you call @mbdavid, but I think he does not have time, so 

 > In our use-case Litedb is feasible or not.
it will work, no issue with that, but you should add more details, eg the possibility for interaction between database, and the use case of what you try to achieve

> Is there any limit to store an amount of data for each individual DB?
as I know, from tests No, unless you exceed the size of the block in your storage driver

> Let me provide if any sample code is available to fit with our use-case.
i do not understand what you means

**Comment by JKamsker** [comment]
Yes, one file per user works well with LiteDB and is a common pattern. Each file is a completely independent database with its own cache, lock and log file. That also makes backup, deletion and per-user encryption (`Password=` in the connection string) easy.

**Limits:** there's no fixed maximum file size; the optional `LimitSize` setting defaults to unlimited. The practical limits are disk space and the ~16 MB maximum size of a single document.

**The one rule:** keep **one** open `LiteDatabase` per file per process and share it, instead of opening a new one per request. A small cache handles that:

```csharp
public sealed class UserDatabases : IDisposable
{
    private readonly string _folder;
    private readonly ConcurrentDictionary<string, Lazy<LiteDatabase>> _open = new();

    public UserDatabases(string folder)
    {
        _folder = folder;
        Directory.CreateDirectory(folder);
    }

    // validate/sanitize userId before it becomes part of a file name
    public LiteDatabase For(string userId) =>
        _open.GetOrAdd(userId, id => new Lazy<LiteDatabase>(
            () => new LiteDatabase(new ConnectionString { Filename = Path.Combine(_folder, id + ".db") }))).Value;

    public void Dispose()
    {
        foreach (var db in _open.Values) db.Value.Dispose();
        _open.Clear();
    }
}

// usage (register UserDatabases as a singleton)
dbs.For(userId).GetCollection<Note>("notes").Insert(new Note { Text = "hello" });
```

It's safe to call from many threads. With a lot of users you may want to dispose the databases of inactive users. If several processes must open the same user's file, use `Connection=Shared` (slower).

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
